### PR TITLE
bugfix(pipeline): Fix state machine of hash join

### LINKED ISSUE #5234: Bug: `COUNT(*)` produce wrong result with joins in new planner
```sql
MySQL [(none)]> select count(*) from numbers(1000) as t inner join numbers(1000) as t1 on t.number = t1.number;
Query OK, 0 rows affected (0.139 sec)
Read 2000 rows, 15.62 KiB in 0.110 sec., 18.19 thousand rows/sec., 142.11 KiB/sec.
```

The expected result is:
```
+----------+
| count(*) |
+----------+
|     1000 |
+----------+
```

While the query `select t.number from numbers(1000) as t inner join numbers(1000) as t1 on t.number = t1.number;` works well.

### PR DESCRIPTION
**Pull request by leiysky**
I hereby agree to the terms of the CLA available at: https://databend.rs/dev/policies/cla/

## Summary

Fix the bug that hash join probe transform won't emit `Finished` event.

## Changelog

- Bug Fix

## Related Issues

Close #5234

### CODE REVIEW

**Comment by BohuTANG** [comment]
LGTM.
Do we have any plan to move some logic to some files from `SQL/exec/mod.rs`? The codes in it are getting more and more trivial :D

**Comment by leiysky** [comment]
> LGTM. Do we have any plan to move some logic to some files from `sql/exec/mod.rs`? The codes in it are getting more and more trivial :D

It's just somewhat glue code. I'm going to do some decouple work on the concepts `PlanNode`, `Expression` and `Transform` after migrating to new planner. After that, we will deal with these files.

**Review by sundy-li — APPROVED** [review]

File: tests/suites/0_stateless/20+_others/20_0001_planner_v2.sql [hunk 1]
```diff
@@ -65,4 +65,6 @@ drop table t;
 drop table t1;
 drop table t2;
 
+select count(*) from numbers(1000) as t inner join numbers(1000) as t1 on t.number = t1.number;
```
Comment (BohuTANG) [inline tests/suites/0_stateless/20+_others/20_0001_planner_v2.sql:68]: Not sure why this is not working:
```
mysql> select * from numbers(1000) as t inner join numbers(1000) as t1 on t.number = t1.number;
ERROR 1105 (HY000): Code: 1002, displayText = Duplicate alias name :number (while in processor thread 0).
```

Comment (BohuTANG) [reply tests/suites/0_stateless/20+_others/20_0001_planner_v2.sql:68]: On, this is OK:
```
mysql> select t.number from numbers(1000) as t inner join numbers(1000) as t1 on t.number = t1.number;
```

Comment (sundy-li) [reply tests/suites/0_stateless/20+_others/20_0001_planner_v2.sql:68]: same problem in #5236

**Pull request merged**